signer/src/stacks/contracts.rs

```diff
@@ -518,20 +531,226 @@ impl AsContractCall for AcceptWithdrawalV1 {
+        if tx_out.value.to_sat() != request.amount {
+            return Err(WithdrawalErrorMsg::InvalidAmount.into_error(ctx, self));
+        }
+        // 7. Check that the fee is less than the desired max-fee.
+        //
+        // The smart contract cannot check if we exceed the max fee, so we
+        // do a check ourselves.
+        //
+        // TODO(552): The better check is to compute what the fee should be
+        // and verify that it matches.
+        if self.tx_fee > request.max_fee {
+            return Err(WithdrawalErrorMsg::InvalidFee.into_error(ctx, self));
+        }
+        // 8. That the signer bitmap matches the bitmap formed from our
+        //    records.
+        let votes = db
+            .get_withdrawal_request_signer_votes(&request.qualified_id(), &ctx.aggregate_key)
+            .await?;
+        if self.signer_bitmap != BitArray::from(votes) {
+            return Err(WithdrawalErrorMsg::BitmapMismatch.into_error(ctx, self));
+        }
+
         Ok(())
     }
+    /// This function validates the sweep transaction.
+    ///
+    /// Specifically, this function:
+    /// 3. That the signer bitcoin transaction sweeping out the users'
+    ///    funds is on the canonical bitcoin blockchain.
+    /// 4. That the sweep transaction has the UTXO indicated by the
+    ///    outpoint.
+    async fn validate_sweep<S>(&self, db: &S, ctx: &ReqContext) -> Result<TxOut, Error>
+    where
+        S: DbRead + Send + Sync,
+    {
+        // First we check that we have a record of the transaction.
+        let txid = self.outpoint.txid.into();
+        let sweep_tx = db
+            .get_bitcoin_tx(&txid, &self.sweep_block_hash)
+            .await?
+            .ok_or_else(|| WithdrawalErrorMsg::SweepTransactionMissing.into_error(ctx, self))?;
+        // 3. That the signer bitcoin transaction sweeping out the users'
+        //    funds is on the canonical bitcoin blockchain.
+        //
+        // From the above, we know that the sweep transaction is in the
+        // `sweep_block_hash`. Now we just need to check that this block is
+        // on the canonical bitcoin blockchain.
+        let block_ref = BitcoinBlockRef {
+            block_hash: self.sweep_block_hash,
+            block_height: self.sweep_block_height,
+        };
+
+        let in_canonical_bitcoin_blockchain = db
+            .in_canonical_bitcoin_blockchain(&ctx.chain_tip, &block_ref)
+            .await?;
+        if !in_canonical_bitcoin_blockchain {
+            return Err(WithdrawalErrorMsg::SweepTransactionReorged.into_error(ctx, self));
+        }
+        // 4. That the sweep transaction has the UTXO indicated by the
+        //    outpoint.
+        //
+        // Okay great, we know that the sweep transaction exists on the
+        // canonical bitcoin blockchain, we just need to do a simple check
+        // of the transaction inputs.
+        sweep_tx
+            .output
+            .get(self.outpoint.vout as usize)
```

Comment:
[Question] Maybe a dumb Q, but is this check sufficient? I mean simply ensuring that the tx contains an output at the expected index? Does that output need to be checked somehow?

Reply:
We check that it is being spent to the intended recipient and that it has the correct amount. Both should be checked, and if they are wrong then we did something wrong or someone is acting maliciously.

Reply:
Ahh right this tx is returned and passed into validate_utxo which does that. Thx, was reading top-down and got mixed up 😅